libvmi/driver/kvm/kvm.c

```diff
@@ -859,22 +858,60 @@ status_t
 kvm_pause_vm(
     vmi_instance_t vmi)
 {
+    unsigned int expect_pause_count = 0;
+
     kvm_instance_t *kvm = kvm_get_instance(vmi);
 
-    if (-1 == kvm->libvirt.virDomainSuspend(kvm->dom)) {
+    // pause vcpus
+    if (kvmi_pause_all_vcpus(kvm->kvmi_dom, &expect_pause_count)) {
+        dbprint(VMI_DEBUG_KVM, "--Failed to pause domain\n");
         return VMI_FAILURE;
     }
+
+    // wait to receive pause events
+    unsigned int pause_events_count = 0;
+    while (pause_events_count < expect_pause_count) {
```

Comment:
Should we simply count the number of pause events received ?
Or also check that a pause events is not sent twice for the same VCPU ?

Reply:
No and no. `kvmi_pause_all_vcpus()` is enough. You don't have to wait the pause events.